### ISSUE TITLE
H::R::Buffer could profit from having an easy "fill with pseudorandom" method

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#3445]
I've found an awful lot of handrolled variants on this both inside Halide and inside Google; would be sensible at this point to consider adding a `fill()` variant that makes this easy (or, perhaps, an adapter that could be passed to the functor variant of `fill()`).

**Comment by dsharletg** [comment]
Isn't this already a one-liner?

    buf.fill([](int, int, int) { return rand(); });

**Comment by steven-johnson** [comment]
Yes, but you need a known-PRNG with a known-seed for many cases.

**Comment by abadams** [comment]
Ideally you could pass a std::uniform_real_distribution to fill() as the callable, but buf.fill has a gotcha. Because it's based on for_each_element, if you don't provide enough args to the lambda, then you only get called on a lower-dimensional slice of the data. This is an intentional feature of for_each_element, but it doesn't work well for fill.

I've been preferring:
```
buf.for_each_value([](float &f) {f = my_rng();});
```

**Comment by steven-johnson** [comment]
Allowing for filling just a lower-dimensional slice of the data as the default action seems like an odd case to optimize for; I can see it being useful sometimes, but having the "wrong" number of args quietly do this seems like a misfeature. What was the driving force for that decision?

**Comment by abadams** [comment]
It's a misfeature for fill, but the intended use for for_each_element is:

```
color_buf.for_each_element([&](int x, int y) {
   color_buf(x, y, 0) = red;
   color_buf(x, y, 1) = green;
   color_buf(x, y, 2) = blue;
});
```

Not sure if anyone actually uses that. We could cut it, and force people to write the following instead:
```
color_buf.sliced(2, 0).for_each_element([&](int x, int y) {
   // same body
}
```

**Comment by dsharletg** [comment]
I use that for fill because it makes it easy to fill an image with particular colors.

That said, I like having to slice explicitly first. The only ugly thing is that it forces you to pick an arbitrary value of the dimension you want to slice, when it doesn't matter.

**Comment by steven-johnson** [comment]
`sliced()` is pretty cheap. I'd be inclined to require that construction and have `for_each_element()` require a compile-time check for the right number of args in the lambda.

**Comment by abadams** [comment]
sgtm. I think it's just a matter of changing assert(dims >= args) to assert(dims == args) in `for_each_element(double, ...)`. That's the overload that handles callables that take some number of ints. Then I guess you'd have to add an overload for fill with a sfinae check for `decltype(std::declval<Fn>()())` to catch cases where the callable takes no args. That should be enough to pass a std distribution to fill.

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the
> reason below. The reasoning was checked against the current source, and a second pass
> tried and failed to refute it, but no human verified this specific text.
> **If the reason is wrong, please reopen the issue and say so** — some fraction of these
> are expected to be wrong.

**Reason (CLOSE_FIXED):** Both halves of the plan agreed in the thread were implemented: `for_each_element` now hard-asserts that the callable's arity equals the buffer dimensionality, and `fill()` has a nullary-callable path that evaluates the callable once per element. `buf.fill([]{ return rand(); })` is now the requested one-liner.

**Evidence:** Commit 5aa42d50f "Assert callable arg count matches buffer dims in for_each_element (#9214)". Current code: src/runtime/HalideBuffer.h:2464-2470 (`assert(args == dims && ...)`) and :2578-2588 (`if constexpr (std::is_invocable_v<Fn &>) { set_host_dirty(); for_each_value([&](T &v){ v = f(); }); }`). Used in anger at test/correctness/interleave_rgb.cpp:84 — `input_buf.fill([]() { return rand(); });`.

Closing as fixed by 5aa42d50f (#9214): `fill()` now accepts a nullary callable and evaluates it once per element, so `buf.fill([&]{ return my_rng(); })` (or any seeded distribution) works, and `for_each_element` now asserts the callable's arity matches the buffer's dimensionality rather than silently filling a slice. See src/runtime/HalideBuffer.h:2464 and :2578, and the usage at test/correctness/interleave_rgb.cpp:84.
